# Telco Customer Churn - Model Evaluation

## Import packages

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, classification_report
from joblib import load

## Load data

In [2]:
df = pd.read_csv('../dataset/dataset_model.csv')

In [3]:
df.describe()

gender  SeniorCitizen      Partner   Dependents       tenure  \
count  7032.000000    7032.000000  7032.000000  7032.000000  7032.000000   
mean      0.504693       0.162400     0.482509     0.298493    32.421786   
std       0.500014       0.368844     0.499729     0.457629    24.545260   
min       0.000000       0.000000     0.000000     0.000000     1.000000   
25%       0.000000       0.000000     0.000000     0.000000     9.000000   
50%       1.000000       0.000000     0.000000     0.000000    29.000000   
75%       1.000000       0.000000     1.000000     1.000000    55.000000   
max       1.000000       1.000000     1.000000     1.000000    72.000000   

       PhoneService  PaperlessBilling  MonthlyCharges  TotalCharges  \
count   7032.000000       7032.000000     7032.000000   7032.000000   
mean       0.903299          0.592719       64.798208   2283.300441   
std        0.295571          0.491363       30.085974   2266.771362   
min        0.000000          0.000000       18.250000     18.800000   
25%        1.000000          0.000000       35.587500    401.450000   
50%        1.000000          1.000000       70.350000   1397.475000   
75%        1.000000          1.000000       89.862500   3794.737500   
max        1.000000          1.000000      118.750000   8684.800000   

             Churn  ...  Contract_Month-to-month  Contract_One year  \
count  7032.000000  ...              7032.000000        7032.000000   
mean      0.265785  ...                 0.551052           0.209329   
std       0.441782  ...                 0.497422           0.406858   
min       0.000000  ...                 0.000000           0.000000   
25%       0.000000  ...                 0.000000           0.000000   
50%       0.000000  ...                 1.000000           0.000000   
75%       1.000000  ...                 1.000000           0.000000   
max       1.000000  ...                 1.000000           1.000000   

       Contract_Two year  PaymentMethod_Bank transfer (automatic)  \
count        7032.000000                              7032.000000   
mean            0.239619                                 0.219283   
std             0.426881                                 0.413790   
min             0.000000                                 0.000000   
25%             0.000000                                 0.000000   
50%             0.000000                                 0.000000   
75%             0.000000                                 0.000000   
max             1.000000                                 1.000000   

       PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
count                            7032.000000                     7032.000000   
mean                                0.216297                        0.336320   
std                                 0.411748                        0.472483   
min                                 0.000000                        0.000000   
25%                                 0.000000                        0.000000   
50%                                 0.000000                        0.000000   
75%                                 0.000000                        1.000000   
max                                 1.000000                        1.000000   

       PaymentMethod_Mailed check  tenure_scaled  MonthlyCharges_scaled  \
count                 7032.000000   7.032000e+03           7.032000e+03   
mean                     0.228100  -1.388410e-16           8.828736e-17   
std                      0.419637   1.000071e+00           1.000071e+00   
min                      0.000000  -1.280248e+00          -1.547283e+00   
25%                      0.000000  -9.542963e-01          -9.709769e-01   
50%                      0.000000  -1.394171e-01           1.845440e-01   
75%                      0.000000   9.199259e-01           8.331482e-01   
max                      1.000000   1.612573e+00           1.793381e+00   

       TotalCharges_scaled  

## Data splitting - training, testing
To be consistent, data will be split using the same random seed and partition sizes as performed in the model training step.

In [4]:
X = df.drop(['Churn'],axis=1) 
y = df['Churn']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=4321, stratify=y)

In [5]:
# raw numerical features are removed from the training, validation and test datasets
X_test = X_test.drop(['tenure','MonthlyCharges','TotalCharges'],axis=1) 

## Load model

In [6]:
model = load('../models/customer_churn_model.joblib') 

## Model evaluation on the test dataset

In [7]:
y_pred = model.predict(X_test)

print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[763 270]
 [ 82 292]]
              precision    recall  f1-score   support

           0       0.90      0.74      0.81      1033
           1       0.52      0.78      0.62       374

    accuracy                           0.75      1407
   macro avg       0.71      0.76      0.72      1407
weighted avg       0.80      0.75      0.76      1407



The model produced a recall of 0.78 and f1-score of 0.62 for positive churn observations, and an overall weight f1-score of 0.76.